### DISCUSSION TITLE
Asking to much from MicroPython?  Get method name from method pointer?

### DISCUSSION THREAD

**Discussion by Jibun-no-Kage** [micropython/micropython#16310] [General]
Asking to much from MicroPython?  Maybe?  Get method name from method pointer?

Using (regular) Python, I often log method calls before the actual call is executed, getting the name of the method from its functional pointer is not always straight forward even in (regular) Python, but is the same in concept doable in MicroPython?  A bridge too far?

**Comment by corranwebster** [comment]
Do you mean something like this?
``` python
def fn():
    ...

print(f"Calling {fn.__name__}")
fn()
```
Which will print `Calling fn` (or instead, just `f"Calling {fn}"` which will give something like `Calling <function fn at 0x20010b20>` for functions, but the much less useful `Calling <bound_method>` for methods).

**Reply by mattytrentini** [reply]
Yes, except I'm not sure what you mean about bound methods? `__name__` works there too (tested with unix v1.24.1):

```python
class Test:
    def foobar(self):
        pass
```
```python
>>> test = Test()
>>> print(test.foobar.__name__)
foobar
```

**Reply by corranwebster** [reply]
The comment about bound methods referred to the `repr(fn)`/`str(fn)`/`f"{fn}"` version without using `__name__`.  For some reason MicroPython doesn't insert the name into the repr of a bound method.

**Reply by mattytrentini** [reply]
Ahh, I see. I wonder how hard it'd be to make the implementation match? Or if there was a significant cost to doing so?

Quick test for those wondering how they differ:

```python
class Foo:
    def bar(self):
        pass

def foobar():
    pass

obj = Foo()

repr(foobar)
repr(obj.bar)
```

Output:

```bash
# MicroPython v1.24.1
'<function foobar at 0x7fdfd02dc3c0>'
'<bound_method 7fdfd02dc620 <Foo object at 7fdfd02dc400>.<function bar at 0x7fdfd02dc3e0>>'
#CPython v3.12
'<function foobar at 0x7facd3983100>'
'<bound method Foo.bar of <__main__.Foo object at 0x7facd3b20380>>'
```

**Reply by mattytrentini** [reply]
FWIW, it looks relatively straightforward to update the repr output - and doesn't look like it'll be costly. Code is in [py/objboundmeth.c#L39](https://github.com/micropython/micropython/blob/master/py/objboundmeth.c#L39) if anyone wants to have a go! 😜

**Comment by bidrohini** [comment]
You can use the string representation of the function or method. 
`class MyClass:
    def my_method(self):
        pass

obj = MyClass()
print(str(obj.my_method))  `
You can parse this string to extract the method name.

**Reply by mattytrentini** [reply]
It's much easier to use `__name__`. See above.

**Comment by Jibun-no-Kage** [comment]
For my purposes... a bit of enhanced tracking/logging __name__ was sufficient.  I had base class, and I had several classes using the same base, but I wanted to log the name of derived class, from within a base class method, for example in __init__ and my 'dispose' (versus __del__ given it is not consistently called automatically in MP) base method, so logging would show when a class was constructed and destructed.
